src/megatron/energon/watchdog.py

```diff
@@ -0,0 +1,291 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import inspect
+import linecache
+import os
+import sys
+import threading
+import time
+import traceback
+from time import perf_counter
+from typing import Any, Callable, Iterable, Iterator, Optional, TypeVar
+
+T = TypeVar("T")  # for the watch_iter type
+
+
+class Watchdog:
+    """
+    A watchdog timer that:
+      - can be 'enabled' or 'disabled' by presence/absence of a deadline,
+      - resets automatically when 'enable()' is called,
+      - can be used as a context manager,
+      - can wrap an iterator to watch only the time for 'next()' calls,
+      - attempts a two-phase shutdown on callback error:
+         1) sys.exit(1) for graceful,
+         2) if still alive after 10s, os._exit(1).
+    """
+
+    def __init__(
+        self,
+        timeout: float,
+        callback: Optional[Callable[[], None]] = None,
+        dump_stacks: bool = True,
+        enabled: bool = True,
+    ) -> None:
+        """
+        :param timeout: Number of seconds before the watchdog fires if not reset/disabled.
+        :param callback: Optional function to call upon timeout.
+        :param dump_stacks: If True, print full stack traces for all threads on timeout (except watchdog's own thread).
+        :param enabled: If False, watchdog starts disabled until enable() is called.
+        """
+        self._timeout = timeout
+        self._callback = callback
+        self._dump_stacks = dump_stacks
+
+        # If _deadline is None, the watchdog is disabled.
+        # Otherwise, _deadline = time.time() + _timeout if enabled.
+        self._deadline: Optional[float] = perf_counter() + timeout if enabled else None
+
+        self._stop = False  # signals permanent shutdown (finish)
+
+        # Condition variable to manage state changes
+        self._cv = threading.Condition()
+        # Background thread (daemon) that monitors timeouts
+        self._worker_thread = threading.Thread(target=self._worker, daemon=True)
+        self._worker_thread.start()
+
+    def _worker(self) -> None:
+        """
+        Background thread that periodically checks if the watchdog has expired.
+        Once it times out or is told to stop, it exits.
+        """
+        while True:
+            with self._cv:
+                if self._stop:
+                    # finish() was called; end the worker.
+                    return
+
+                if self._deadline is None:
+                    # Disabled; no deadline. Just wait a bit, then re-check.
+                    self._cv.wait(timeout=1.0)
+                    continue
+
+                remaining = self._deadline - perf_counter()
```

Comment:
Missing a lock. The main thread my `reset` or `disable` / `finish` here, which may make `_deadline = None`, i.e. fail here.